src/utils/localStorage.ts

```diff
@@ -0,0 +1,74 @@
+import { Result } from "types/functional";
+import { Err, Ok } from "./functional";
+
+/**
+ * Gets data from a key in local storage, parses it into its expected type, and returns it.
+ * If no data is found, or the data is invalid, undefined is returned instead.
+ * @param key - The key in Local Storage to fetch from
+ * @param expectedType - The type expected from Local Storage
+ * @returns The stored data in its expected type, or undefined if not found or valid.
+ */
+export function getLocalStorage<T>(key: string, expectedType: string): Result<T, string> {
+  const rawLocalStorage = window.localStorage.getItem(key) ?? undefined;
+  if (rawLocalStorage === undefined) return Err(`LocalStorage did not contain an entry for key ${key}`);
+
+  switch (expectedType) {
+    case "string":
+      return Ok(rawLocalStorage as T);
+    case "number":
+      return Ok(+rawLocalStorage as T);
+    case "boolean":
+      return Ok((rawLocalStorage.trim().toLowerCase() === "true") as T);
+    case "object":
+      try {
+        const parsedJson = JSON.parse(rawLocalStorage);
+        return Ok(parsedJson);
+      } catch (why: unknown) {
+        return Err("JSON object could not be parsed from LocalStorage");
+      }
+    default:
+      return Err(`Type of ${expectedType} is not supported by useLocalStorage.getLocalStorage()`);
+  }
+}
```

Comment:
Is there a way for typescript to check if the "expectedType" parameter actually matches the type parameter T at compile time? We also do a typeof check on rawLocalStorage to make sure it matches the expected type string passed in before returning Ok().